--- /Users/sneridagh/Development/plone/volto/node_modules/razzle/config/createJestConfig.js	2020-11-07 16:11:14.489360747 +0100
+++ /Users/sneridagh/Development/plone/volto/node_modules/razzle/config/createJestConfigNew.js	2020-11-07 16:13:44.816817775 +0100
@@ -84,6 +84,7 @@
       'mapCoverage',
       'moduleDirectories',
       'moduleFileExtensions',
+      'modulePathIgnorePatterns',
       'moduleNameMapper',
       'modulePaths',
       'snapshotSerializers',
